ql/src/java/org/apache/hadoop/hive/ql/stats/BasicStatsTask.java

```diff
@@ -512,6 +513,8 @@ private String toString(Map<String, String> parameters) {
         if (dpPartSpecs != null) {
           // load the list of DP partitions and return the list of partition specs
           list.addAll(dpPartSpecs);
+          // Reload partition metadata because another BasicStatsTask instance may have updated the stats.
+          list = db.getPartitionsByNames(table, list.stream().map(Partition::getName).collect(Collectors.toList()));
```

Comment:
This change is kind of reverting https://issues.apache.org/jira/browse/HIVE-15250. Obviously correctness is more important than performance but I am wondering if we explored other ways to fix the problem. 

@rbalamohan since you worked on HIVE-15250 you may want to have a look into the changes here.

Reply:
Truth is that things are a bit different here (in comparison with the state before HIVE-15250) in the sense that we are calling `getPartitionsByNames` instead of `getPartition` so if I understand well it will be only one call to DB and not multiple so maybe the perf overhead is not as important as it was in HIVE-15250.

Reply:
I think these are thrift API calls from HS2 to HMS and the number of calls depend on the number of partitions and `hive.metastore.batch.retrieve.max`.
Also there is some communication from HMS to backend db which is also done in batches.

I explored another way:
* In the StatsTask instead of applying the stats on the Partition objects collect it to a delta objects.
* Implement a HMS call and backend DB call to update stats based on the collected delta objects.
However this would be a bigger change.

Reply:
getPartitionByNames is a single call unlike 2000+ getPartition() calls (Which internally loads table object everytime etc). It will still add up to runtime depending on the number of partitions for regular insert queries.

One option could be to figure out if it is multi-insert at the time of sem analyzer and pass it on to DynamicPartitionCtx; getPartitionByNames() could be invoked only in those conditions.

Reply:
Given that we are dealing with a correctness problem here, I would be inclined to commit this fix as is right now and treat possible improvements if any in a follow up keeping in mind that the performance overhead of `getPartitionByNames` is not measured.